### PR TITLE
add service address when register service

### PR DESCRIPTION
**Pull request by kopinions**

### CODE REVIEW

**Comment by robynbergeron** [comment]
@sjkyspa  -- thanks for this pull request.

@sgargan -- Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.
If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by sgargan** [comment]
@sjkyspa. I'm not sure that the address is actually required for a service registration, it just defaults to the agent host. If this is the case the documentation could do with updating to reflect that this is optional.

There are tests for the consul support in the core project where the rest of the tests reside, wouldn't hurt to add a test for this.

Thanks for taking the time to add this.

**Comment by kopinions** [comment]
@sgargan  I think the address is required for service registration. Without the address, the service registered to the consul will left the serivce_address blank. I will not be discoverable to other app.

**Comment by sgargan** [comment]
@sjkyspa check the docs here https://www.consul.io/docs/agent/services.html the address is optional. If you register a service by calling the module on the host where the agent is running it will use the address of that agent. If you register the service using a local action then yes you'll need to supply the address.

**Comment by kopinions** [comment]
@robynbergeron 
when will the pull request be merged to the master?

**Comment by robynbergeron** [comment]
Hi @sjkyspa  -- 

@sgargan is the maintainer of this module -- it's not entirely clear to me if he wants changes, but it looks like he was hoping to have the details around address requirements for service registration clarified. 

Once we have response from @sgargan saying "shipit", then we can mark it for inclusion -- but it would also be helpful to get a "needs_revision" comment, if @sgargan thinks revision is needed :)

Thanks!

**Comment by kopinions** [comment]
@sgargan The address is optional, that right. But when you left the address empty, then the serivce registered with left the serivce address as empty. The normal is showed blow
image

**Comment by logic855** [comment]
Hi, when will this pull request be merged? I could really use this nice feature.

**Comment by sgargan** [comment]
Sorry I'd not seen the updates for this :ship:

**Comment by kopinions** [comment]
@sgargan Do you still need me to do something? Feel free to tell me.

**Comment by sgargan** [comment]
No, I'm happy, I'd commented it should ship. shipit

**Comment by gregdek** [comment]
Thanks again to @sjkyspa for this PR, and thanks @sgargan for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: clustering/consul.py [hunk 1]
```diff
@@ -81,6 +81,11 @@
           - the port on which the service is listening required for
             registration of a service, i.e. if service_name or service_id is set
         required: false
+    service_address:
+        description:
+          - the address on which the service is serving required for
+            registration of a service
+        required: false
```
Comment (resmo) [inline clustering/consul.py:88]: please add the following to the new param docs:

```
default: localhost
version_added: "2.1"
```

**Comment by joelthompson** [comment]
I think the default value of "localhost" is broken, and is certainly a breaking change. I believe it should be `None`.

If you do that, then the service will register as advertising itself as being at localhost, even when you query from another box. The default should be None. Even though the address shown when queried from the [`/v1/agent/services`](https://www.consul.io/docs/agent/http/agent.html#agent_services) endpoint is empty, Consul takes care of that for you automatically when responding to queries.

For example, I registered `myservice` using the default `service_address`, and then when I run this query on my Consul server itself (which is a different machine):

```
$ dig @localhost -p 8600 myservice.service.consul

; <<>> DiG 9.9.4-RedHat-9.9.4-29.el7_2.3 <<>> @localhost -p 8600 myservice.service.consul
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 6905
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;myservice.service.consul. IN A

;; ANSWER SECTION:
myservice.service.consul. 0 IN CNAME localhost.

;; Query time: 490 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Fri Jun 17 04:56:28 UTC 2016
;; MSG SIZE  rcvd: 125
```

(Note that localhost doesn't get resolved properly, which looks similar to hashicorp/consul#1228) localhost is the wrong value there. When I change the default to `None` in the module, it works as expected.

I'll try to get a PR submitted to fix this.

**Pull request merged**